src/NServiceBus.TransactionalSession/PipelineAwareTransactionalSession.cs

```diff
@@ -0,0 +1,160 @@
+            }
+        }
+
+        public string SessionId
+        {
+            get
+            {
+                if (!IsOpen)
+                {
+                    throw new InvalidOperationException(
+                        "The session has to be opened before accessing the SessionId.");
+                }
+
+                return pipelineContext!.MessageId;
+            }
+        }
+
+        // In the invoke handler context phase the synchronized storage is ready to use
+        bool IsOpen => pipelineContext is not null;
+
+        public Task Open(OpenSessionOptions options, CancellationToken cancellationToken = default)
+        {
+            ThrowIfDisposed();
+            ThrowIfCommitted();
+
+            if (IsOpen)
+            {
+                throw new InvalidOperationException($"This session is already open. {nameof(ITransactionalSession)}.{nameof(ITransactionalSession.Open)} should only be called once.");
+            }
+
+            pipelineContext = ((PipelineAwareSessionOptions)options).PipelineContext;
+
+            foreach (var customization in customizations)
+            {
+                customization.Apply(options);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public Task Send(object message, SendOptions sendOptions, CancellationToken cancellationToken = default)
+        {
+            ThrowIfInvalidState();
+
+            cancellationToken.ThrowIfCancellationRequested();
+            // unfortunately there is no way to marry both tokens together
+            return pipelineContext!.Send(message, sendOptions);
+        }
+
+        public Task Send<T>(Action<T> messageConstructor, SendOptions sendOptions, CancellationToken cancellationToken = default)
+        {
+            ThrowIfInvalidState();
+
+            cancellationToken.ThrowIfCancellationRequested();
+            // unfortunately there is no way to marry both tokens together
+            return pipelineContext!.Send(messageConstructor, sendOptions);
+        }
+
+        public Task Publish(object message, PublishOptions publishOptions, CancellationToken cancellationToken = default)
+        {
+            ThrowIfInvalidState();
+
+            cancellationToken.ThrowIfCancellationRequested();
+            // unfortunately there is no way to marry both tokens together
+            return pipelineContext!.Publish(message, publishOptions);
+        }
+
+        public Task Publish<T>(Action<T> messageConstructor, PublishOptions publishOptions,
+            CancellationToken cancellationToken = default)
+        {
+            ThrowIfInvalidState();
+
+            cancellationToken.ThrowIfCancellationRequested();
+            // unfortunately there is no way to marry both tokens together
+            return pipelineContext!.Publish(messageConstructor, publishOptions);
+        }
+
+        public Task Commit(CancellationToken cancellationToken = default)
```

Comment:
> so when using the pipeline aware session, users shouldn't call commit because that would then throw when the pipeline behavior tries to commit the session. I'm kinda wondering whether this is clear enough to the user (since this code isn't visible to them at all).

The thing is in the near future, the majority of the lifetime management will be moved into some sort of filter pipeline things even for the controllers. So in essence for all those scenarios the user will no longer manage the session lifetime anyway. This just does that already by implementing the automatic management of the session in the pipeline.

Happy to change the implementation though if you prefer that to allow multiple commits. I do prefer though the current way because it would make discover invalid usages early on during development